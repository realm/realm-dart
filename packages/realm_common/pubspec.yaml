name: realm_common
description: >-
             Hosts the common code shared between realm, realm_dart and realm_generator packages.
             This package is part of the official Realm Flutter and Realm Dart SDKs.

version: 3.2.0

homepage: https://www.realm.io
repository: https://github.com/realm/realm-dart
issue_tracker: https://github.com/realm/realm-dart/issues

environment:
  sdk: ^3.3.0

dependencies:
  collection: ^1.18.0
  objectid: ^3.0.0
  sane_uuid: ^1.0.0

dev_dependencies:
  lints: ^3.0.0
