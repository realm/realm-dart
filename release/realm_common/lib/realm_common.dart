// Copyright 2021 MongoDB, Inc.
// SPDX-License-Identifier: Apache-2.0

export 'src/realm_common_base.dart';
export 'src/realm_types.dart';

export 'package:objectid/objectid.dart' show ObjectId;
export 'package:sane_uuid/uuid.dart' show Uuid;
