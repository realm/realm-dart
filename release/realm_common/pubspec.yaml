name: realm_common
description: >-
             Hosts the common code shared between realm, realm_dart and realm_generator packages.
             This package is part of the official Realm Flutter and Realm Dart SDKs.

version: 2.0.0-alpha.5

homepage: https://www.realm.io
repository: https://github.com/realm/realm-dart
issue_tracker: https://github.com/realm/realm-dart/issues

environment:
  sdk: ^3.0.0

dependencies:
  collection: ^1.18.0
  objectid: ^3.0.0
  sane_uuid: ^1.0.0-alpha.5

dev_dependencies:
  lints: ^3.0.0
