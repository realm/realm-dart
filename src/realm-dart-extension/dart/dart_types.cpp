////////////////////////////////////////////////////////////////////////////
//
// Copyright 2016 Realm Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////
#pragma clang diagnostic ignored "-Wdocumentation"

#include <cmath>
#include <functional>
#include <map>
#include <string>

#include "dart_api.h"
#include "dart_native_api.h"

#include "js_types.hpp"

#include "dart_types.hpp"

bool ___DartError::use_native_exceptions;

Dart_Handle operator ||(const Dart_Handle& dartResult, ___DartError error) {
	if (Dart_IsError(dartResult)) {
		auto err = Dart_GetError(dartResult);
		//Dart_PropagateError(dartResult);
		if (!___DartError::use_native_exceptions) {
			Dart_ThrowException(dartResult);
		}
		else {
			throw std::runtime_error(err);
		}
		
	}
	return dartResult;
}

namespace realm {
	namespace dartvm {
		namespace flutter {
			std::string FilesDir;
		}

		Dart_Handle DartCoreLibrary;
		Dart_Handle RealmLibrary;
		std::string RealmPackageName;
		Dart_Handle DartDateTimeType;
		
		Dart_Handle RealmCreateDateTimeFunc;
		Dart_Handle RealmDynamicObjectType;
		Dart_Handle RealmTypeStaticPropertiesType;
		Dart_Handle RealmHelpersType;
	}

	namespace util {
		void to_lower(char& c) {
			c = std::tolower(static_cast<unsigned char>(c));
		}

		void to_lowecase(std::string& string)
		{
			std::for_each(string.begin(), string.end(), to_lower);
		}
	}
}

namespace Dart {

	HandleScope::HandleScope() {
		Dart_EnterScope();
	}

	HandleScope::~HandleScope() {
		Dart_ExitScope();
	}

	Dart_Handle GetLibrary(const char* libName) {
		Dart_Handle libraryUrl = Dart_NewStringFromCString(libName) || handleError;
		Dart_Handle library = Dart_LookupLibrary(libraryUrl) || handleError;
		return library;
	}
}

