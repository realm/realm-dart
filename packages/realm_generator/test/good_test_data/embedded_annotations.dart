import 'package:realm_common/realm_common.dart';

part 'embedded_annotations.realm.dart';

@RealmModel()
class _Parent {
  @MapTo('single child')
  late _Child1? child;

  @MapTo('CHILDREN')
  late List<_Child1> children;
}

@RealmModel(ObjectType.embeddedObject)
@MapTo('MySuperChild')
class _Child1 {
  @MapTo('_value')
  late String value;

  @MapTo('_parent')
  late _Parent? linkToParent;

  @Indexed()
  late String indexedString;

  @Ignored()
  late String ignoredString;
}
