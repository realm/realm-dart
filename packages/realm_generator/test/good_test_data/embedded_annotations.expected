// GENERATED CODE - DO NOT MODIFY BY HAND

part of 'embedded_annotations.dart';

// **************************************************************************
// RealmObjectGenerator
// **************************************************************************

// ignore_for_file: type=lint
class Parent extends _Parent with RealmEntity, RealmObjectBase, RealmObject {
  Parent({
    Child1? child,
    Iterable<Child1> children = const [],
  }) {
    RealmObjectBase.set(this, 'single child', child);
    RealmObjectBase.set<RealmList<Child1>>(
        this, 'CHILDREN', RealmList<Child1>(children));
  }

  Parent._();

  @override
  Child1? get child =>
      RealmObjectBase.get<Child1>(this, 'single child') as Child1?;
  @override
  set child(covariant Child1? value) =>
      RealmObjectBase.set(this, 'single child', value);

  @override
  RealmList<Child1> get children =>
      RealmObjectBase.get<Child1>(this, 'CHILDREN') as RealmList<Child1>;
  @override
  set children(covariant RealmList<Child1> value) =>
      throw RealmUnsupportedSetError();

  @override
  Stream<RealmObjectChanges<Parent>> get changes =>
      RealmObjectBase.getChanges<Parent>(this);

  @override
  Parent freeze() => RealmObjectBase.freezeObject<Parent>(this);

  EJsonValue toEJson() {
    return <String, dynamic>{
      'single child': child.toEJson(),
      'CHILDREN': children.toEJson(),
    };
  }

  static EJsonValue _toEJson(Parent value) => value.toEJson();
  static Parent _fromEJson(EJsonValue ejson) {
    return switch (ejson) {
      {
        'single child': EJsonValue child,
        'CHILDREN': EJsonValue children,
      } =>
        Parent(
          child: fromEJson(child),
          children: fromEJson(children),
        ),
      _ => raiseInvalidEJson(ejson),
    };
  }

  static final schema = () {
    RealmObjectBase.registerFactory(Parent._);
    register(_toEJson, _fromEJson);
    return SchemaObject(ObjectType.realmObject, Parent, 'Parent', [
      SchemaProperty('child', RealmPropertyType.object,
          mapTo: 'single child', optional: true, linkTarget: 'MySuperChild'),
      SchemaProperty('children', RealmPropertyType.object,
          mapTo: 'CHILDREN',
          linkTarget: 'MySuperChild',
          collectionType: RealmCollectionType.list),
    ]);
  }();

  @override
  SchemaObject get objectSchema => RealmObjectBase.getSchema(this) ?? schema;
}

class Child1 extends _Child1 with RealmEntity, RealmObjectBase, EmbeddedObject {
  Child1(
    String value,
    String indexedString, {
    Parent? linkToParent,
  }) {
    RealmObjectBase.set(this, '_value', value);
    RealmObjectBase.set(this, '_parent', linkToParent);
    RealmObjectBase.set(this, 'indexedString', indexedString);
  }

  Child1._();

  @override
  String get value => RealmObjectBase.get<String>(this, '_value') as String;
  @override
  set value(String value) => RealmObjectBase.set(this, '_value', value);

  @override
  Parent? get linkToParent =>
      RealmObjectBase.get<Parent>(this, '_parent') as Parent?;
  @override
  set linkToParent(covariant Parent? value) =>
      RealmObjectBase.set(this, '_parent', value);

  @override
  String get indexedString =>
      RealmObjectBase.get<String>(this, 'indexedString') as String;
  @override
  set indexedString(String value) =>
      RealmObjectBase.set(this, 'indexedString', value);

  @override
  Stream<RealmObjectChanges<Child1>> get changes =>
      RealmObjectBase.getChanges<Child1>(this);

  @override
  Child1 freeze() => RealmObjectBase.freezeObject<Child1>(this);

  EJsonValue toEJson() {
    return <String, dynamic>{
      '_value': value.toEJson(),
      '_parent': linkToParent.toEJson(),
      'indexedString': indexedString.toEJson(),
    };
  }

  static EJsonValue _toEJson(Child1 value) => value.toEJson();
  static Child1 _fromEJson(EJsonValue ejson) {
    return switch (ejson) {
      {
        '_value': EJsonValue value,
        '_parent': EJsonValue linkToParent,
        'indexedString': EJsonValue indexedString,
      } =>
        Child1(
          fromEJson(value),
          fromEJson(indexedString),
          linkToParent: fromEJson(linkToParent),
        ),
      _ => raiseInvalidEJson(ejson),
    };
  }

  static final schema = () {
    RealmObjectBase.registerFactory(Child1._);
    register(_toEJson, _fromEJson);
    return SchemaObject(ObjectType.embeddedObject, Child1, 'MySuperChild', [
      SchemaProperty('value', RealmPropertyType.string, mapTo: '_value'),
      SchemaProperty('linkToParent', RealmPropertyType.object,
          mapTo: '_parent', optional: true, linkTarget: 'Parent'),
      SchemaProperty('indexedString', RealmPropertyType.string,
          indexType: RealmIndexType.regular),
    ]);
  }();

  @override
  SchemaObject get objectSchema => RealmObjectBase.getSchema(this) ?? schema;
}
